brille/design/Proposed_integrated_syntax_Brill_SpinW_Horace.md

```diff
@@ -0,0 +1,267 @@
+
+## Extracting the data
+
+For single crystal data, this is currently possible in Matlab using Horace:
+
+```matlab
+proj = projaxes([1 0 0], [0 1 0]);
+cut1 = cut_sqw(sqw_file, proj, [hmin hstep hmax], [kmin kstep kmax], [lmin lmax], [emin estep emax]);
+```
+
+The above example assumes that data processing to construct an `.sqw` file combining multiple measurements of a crystal at different orientations is done.
+Triplets of numbers indicate that the data is to be rebinned along that axis (defined by the projection `proj`), whilst doublets indicates the data should be integrated between those ranges.
+Thus the example generates a 3D `sqw` object `cut1` from the 4D data in the `sqw_file`.
+
+The proposed syntax for Python is:
+
+<table>
+<td><img src="https://placehold.it/5x100/000000/000000?text=+"></td>
+<td>
+<pre lang="python">
+from pyHorace import horace as h
+proj = {'u':[1,0,0], 'v':[0,1,0], 'type':'rrr'}
+cut1 = h.cut_sqw(sqw_file, proj, [hmin, hstep, hmax], [kmin, kstep, kmax], 
+                                 [lmin, lmax], [emin, estep, emax])
+</pre>
+</td>
+<td><img src="https://placehold.it/5x100/000000/000000?text=+"></td>
+</table>
+
+For powder data, this is currently possible using the MSlice program.
+There are two implementations of MSlice: [one](https://github.com/mantidproject/mslice) in Python as part of [Mantid](https://www.mantidproject.org)
+and [the original](http://mslice.isis.rl.ac.uk/Main_Page) written in Matlab.
+The original Matlab MSlice is mostly a GUI program with an undocumented CLI (by using the underlying functions of the GUI).
+This program is also currently not being maintained so we will not use it.
+
+In Python, the syntax to extract data from a powder measurement using Mantid MSlice is:
+
+```python
+import mslice.cli as m
+dat2 = m.Load(nxs_file)
+cut2 = m.Slice(d2, f'|Q|,{qmin},{qmax},{qstep}', f'DeltaE,{emin},{emax},{estep}')
+```
+
+Note that unlike in Horace, the original pixel information is not retained in this rebinning operation.
+
+We propose the following syntax for powder data handling in Horace, should it be implemented:
+
+<table>
+<td><img src="https://placehold.it/5x70/000000/000000?text=+"></td>
+<td>
+<pre lang="matlab">
+proj = projaxes('powder');
+cut2 = h.cut_sqw(sqw_file, proj, [qmin, qstep, qmax], [emin, estep, emax]);
+</pre>
+</td>
+<td><img src="https://placehold.it/5x70/000000/000000?text=+"></td>
+</table>
+
+This syntax should retain pixel information just like the equivalent single crystal invocation.
+
+
+## Modelling the data
+
+We assume that [SpinW](https://spinw.org/) would be used to model the magnetic excitations.
+To set up and evaluate a model on single crystal data with current syntax in Matlab is:
+
+```matlab
+tri = sw_model('triAF', 1); 
+spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
+sim1 = disp2sqw_eval(cut1, @tri.horace_sqw, spinw_pars, fwhm);
```

Comment:
It's an interesting philosophical point, actually: Should *Horace* provide the interface to "user written" model or just provide an API, and *is* SpinW (and Euphonic) a "user written" (i.e. third party) model? I guess I went with the "just provide an API" option and treated SpinW the same as another third party code - the fact that we're all contributing to these codes and they are interconnected just blurs the line a bit... I guess we could discuss this at the meeting.

In fact, Horace does have such [interface functions](https://github.com/pace-neutrons/Horace/blob/master/horace_core/functions/spinw_sqw.m). These functions were actually written first and if I hadn't had any involvement with SpinW it would probably be the way we'd still do it. The issue is that there are certain coordinate transformations and optimisations in the model evaluation steps which needed modifications to the SpinW object itself. We found that ([a](https://github.com/brille/brillem/blob/master/%2Bbrillem/%40Brille/spinwave.m) [bit](https://github.com/brille/brillem/blob/master/%2Bbrillem/%40Brille/neutron_spinwave_intensity.m) [like](https://github.com/brille/brillem/blob/master/%2Bbrillem/%40Brille/sw_sab.m) Brillem we had to repeat some code in Horace or to store Horace information in a blank field in the SpinW object to pass from iteration to iteration).

In the end, because I was a SpinW contributor, it was easier to deal with it in SpinW. This could be a SpinW specific thing though. So the basic usage of SpinW is to define the model with all the exchange parameters, geometry etc, and then to calculate the spin wave dispersion at some set of `hkl`. During fitting we don't want to do all the set up steps, just to change a (sub-)set of exchange parameters (or similar). It is this "performance optimisation" which required changes to the SpinW code. This wouldn't really apply to Euphonic since you're not going to be varying the force constants.